### PR TITLE
Revisiting non-conditional short-circuit handling

### PR DESCRIPTION
**Pull request by sonalmahajan15**
This PR revisits the non-conditional short-circuit handling. Following are the changes:

- The main change is to repurpose the `AddNilCheck` function to trigger the handling, instead of the earlier custom (and duplicated) implementation in `asNilCheckExpr`. 
- `AddNilCheck` is strengthened to also be able to handle negations. E.g., `!(x == nil)`
- The added benefit of using `AddNilCheck()` is also that it already has support for `len()` checks, which the non-conditional handling can leverage right away. A new extensive set of test cases for len checks (as pointed in issue #92) is also added.

### CODE REVIEW

File: assertion/function/assertiontree/util.go [hunk 1]
```diff
@@ -179,10 +179,29 @@ func inverseToken(t token.Token) token.Token {
...
+		// and false branches produced by the binary expression. For example, if`!(v != nil)`, then AddNilCheck on the inner
+		// expression `(v != nil)` returns trueCheck: produceNegativeNilCheck, falseCheck: noop, isNoop = false, implying a
+		// negative nil check for the true branch. But since it is preceded with a negation (!), the below code
+		// interchanges the true and false branches, and returns trueCheck: noop, falseCheck: produceNegativeNilCheck,
+		// implying a negative nil check for the false branch.
+		if e.Op == token.NOT {
+			trueNilCheck, falseNilCheck, isNoop := AddNilCheck(pass, e.X)
+			return falseNilCheck, trueNilCheck, isNoop
+		}
+	}
```
Comment (yuxincs) [inline assertion/function/assertiontree/util.go:200]: nit: seems simpler to get rid of the switch case and just treat the `!(expr)` as a special case and early return, right?

```go
if e, ok := expr.(*ast.UnaryExpr); ok && e.Op == token.NOT {
  return ...
}

binExpr, ok := expr
if !ok {
  return ...
}
```

File: testdata/src/go.uber.org/nilcheck/nonconditionalflow.go [hunk 2]
```diff
@@ -202,3 +202,50 @@ func testChainedAccesses(x *X, i int) bool {
 	}
 	return false
 }
+
+// ---- test len checks ----
```
Comment (yuxincs) [inline testdata/src/go.uber.org/nilcheck/nonconditionalflow.go:206]: Nice addition to the test suite!!

**Review by yuxincs — APPROVED** [review]
Great find and nice tech debt cleanup!

I only have a nit for readability (maybe), feel free to disagree.

File: assertion/function/assertiontree/util.go (hunk 1 above)
Comment (sonalmahajan15) [reply assertion/function/assertiontree/util.go]: Fair enough. Updated the code to replace the `switch`.

**Pull request merged**